lib/MongoDB/ClientSession.pm

```diff
@@ -98,6 +114,34 @@ has _server_session => (
     clearer => '__clear_server_session',
 );
 
+has _current_transaction_settings => (
+    is => 'rwp',
+    isa => HashRef,
+    init_arg => undef,
+    clearer => '_clear_current_transaction_settings',
+);
+
+has _transaction_state => (
+    is => 'rwp',
+    isa => TransactionState,
+    default => 'none',
+);
+
+# Flag used to say we are still in a transaction
+has _active_transaction => (
+    is => 'rwp',
+    isa => Boolish,
+    default => 0,
+);
+
+# Flag used to say whether any operations have been performed on the
+# transaction
+has _has_transaction_operations => (
```

Comment:
Not fully - the biggest problem is being able to call commit multiple times. Its easy if you are coming from starting or in_progress states, but after that you are in the committed state - at which point, its unknown what state you were in before. I thought about tracking previous state as well, but if you then call commit more than twice (which is allowed AFAICT) it would break - so find just setting a boolean more useful.

Reply:
ok.